### ISSUE TITLE
How to send concurrent requests?

### ISSUE THREAD

**Issue by mxr576** [php-http/client-common#105]
Is there a way to send concurrent requests and wait for their results in a same way as [Guzzle supports](http://docs.guzzlephp.org/en/stable/quickstart.html#concurrent-requests)? 
Ex.: Load entities by ids in a function from on API that can only load entities one by one.

(#7 Might be related to this question.)

**Comment by mxr576** [comment]
Okay, it seems Httplug promises and `\GuzzleHttp\Promise\settle()` and `\GuzzleHttp\Promise\unwrap()` are working fine together, so let me rephrase my question. How can my code remain client agnostic if I would like to implement the above mentioned example with another client adapter?  Should I require and use `guzzlehttp/promises` for curl, react and others too?

**Comment by mxr576** [comment]
Tested cUrl adapter, it seems this one also working with Guzzle promises (not so surprisingly). Haven't tested other one.

**Comment by dbu** [comment]
glad you found a solution. the recommended way of doing this is using the HttpAsyncClient with a request, which returns you a php-http/promise . see http://docs.php-http.org/en/latest/components/promise.html

(for background: there should be a PSR - or even better a PHP language level solution - for promises. a PSR once started but so far has not been released, so we had to do our own promises for php-http.)

**Comment by mxr576** [comment]
@dbu Any estimate when the "own promises for php-http" could actually happen? Is there a related issue that I can follow?
